### ISSUE TITLE
Write the witness file for concurrent programs

### ISSUE THREAD

**Issue by lucasccordeiro** [marek-trtik/cbmc#11]
The format for violation witnesses for concurrent tasks is described here https://github.com/sosy-lab/sv-witnesses#additional-edge-data-for-concurrent-programs.

**Comment by tautschnig** [comment]
Am I right that a9e31b64b3 is a fix for this issue?

**Comment by peterschrammel** [comment]
That was a first attempt. Looking at the current results, this requires more work. Currently, I see only two witnesses confirmed. I'm looking into this right now.

**Comment by peterschrammel** [comment]
Stripping the witness produced by 

./cbmc ../sv-benchmarks/c/pthread/fib_bench_false-unreach-call.i --propertyfile ../sv-benchmarks/c/ConcurrencySafety.prp --graphml-witness witness.graphml --32

down to 

[witness_works2.graphml.txt](https://github.com/marek-trtik/cbmc/files/1504448/witness_works2.graphml.txt)

seems to give something acceptable for cpachecker. I've also seen witness by other tools that just have entry and violation nodes... which is also accepted...

**Comment by marek-trtik** [comment]
I am currently working on witness validation of `pthread-wmm/mix000_power.oepc_false-unreach-call.i`.

**Comment by marek-trtik** [comment]
I'll try to fix these 2 issue first:
1. Missing `<data key="specification">CHECK( init(main()), LTL(G ! call(__VERIFIER_error())) )</data>` (CBMC does not write it, while ESBMC does)
2. Strange assumptions `<data key="assumption">\result = FALSE;</data>`

**Comment by tautschnig** [comment]
The specification field is inserted by the wrapper script, so it should really be there?

For the "strange" assumptions: which part of it do you consider "strange"?

**Comment by peterschrammel** [comment]
In my experience the strategy to solve witness problems is to manually modify the witnesses that we produce until they are accepted and only then to start implementing the changes necessary to produce the modified witness.

**Comment by peterschrammel** [comment]
Typical modifications that might lead to success are 
* suppressing assumptions
* suppressing nodes

For example, in my example above, I removed all assumptions and kept only one node per context switch.

**Comment by marek-trtik** [comment]
@tautschnig :
1. The "missing specification" issue is not an issue. It is present. I did not know it is generated in the wrapper script.
2. The condition https://github.com/marek-trtik/cbmc/blob/sv-comp-2018-patches/src/goto-programs/graphml_witness.cpp#L164  does not for benchmarks containing variables with the `$` character. For exmaple, `int y$mem_tmp;` in `pthread-wmm/mix000_power.oepc_false-unreach-call.i`. I fixed that by replacing `$` by `#return_value`, according to description of `assumption.resultfunction` in page https://github.com/sosy-lab/sv-witnesses

**Comment by marek-trtik** [comment]
@peterschrammel : I applied the changes you proposed. And, for example, for `pthread-wmm/mix000_power.oepc_false-unreach-call.i` we now get this result
`Verification result: TRUE. No property violation found by chosen configuration.`
But the correct answer should be:
Verification result: FALSE. Property violation (WitnessAutomaton) found by chosen configuration.

**Comment by tautschnig** [comment]
> I fixed that by replacing `$` by `#return_value`

Will this also work for functions that don't have a body, such as `__VERIFIER_nondet_*`?

**Comment by marek-trtik** [comment]
@tautschnig : If you mean variables of the form `return_value___VERIFIER_nondet_*$N`, like is this GOTO program:
```
        // 168 file ../sv-benchmarks/c/pthread-wmm/mix000_power.oepc_false-unreach-call.i line 781 function P1
        void *return_value___VERIFIER_nondet_pointer$1;
        // 169 file ../sv-benchmarks/c/pthread-wmm/mix000_power.oepc_false-unreach-call.i line 781 function P1
        return_value___VERIFIER_nondet_pointer$1 = NONDET(void *);
        // 170 file ../sv-benchmarks/c/pthread-wmm/mix000_power.oepc_false-unreach-call.i line 781 function P1
        P1#return_value = return_value___VERIFIER_nondet_pointer$1;
```
then no. I can upgrade the condition to capture these variables as well.

**Comment by tautschnig** [comment]
It's ok if either of 169 or 170 results in a `\result = ...` assumption in the witness. It's not necessary to cover both of them, but please do make sure that one of them ends up in there.

**Comment by marek-trtik** [comment]
Thanks for the notes. To be sure we do not miss anything I upgraded the condition to capture both cases.

**Comment by marek-trtik** [comment]
Here is the commit that provides the improved validation of our witnesses:
https://github.com/marek-trtik/cbmc/commit/68fee5b5e032bf73f1e4c97f4f3d6ed5c826f44a

I checked the improvement on these benchmarks:
```
"pthread-wmm/mix000_power.oepc_false-unreach-call.i",     # OK!
"pthread-wmm/mix000_power.opt_false-unreach-call.i",     # OK!
"pthread-wmm/rfi008_tso.opt_false-unreach-call.i",    # OK!
"pthread-lit/fkp2013_variant_false-unreach-call.i",       # PROBLEM! Verification result: UNKNOWN, incomplete analysis.
"ldv-races/race-1_3-join_false-unreach-call.i",       # OK!
"pthread-driver-races/char_generic_nvram_read_nvram_write_nvram_false-unreach-call.i",      # OK!
"pthread-C-DAC/pthread-demo-datarace_false-unreach-call.i",           # OK!
"pthread/lazy01_false-unreach-call.i",        # OK!
"pthread/stack_false-unreach-call.i",     # PROBLEM! Verification result: UNKNOWN, incomplete analysis. Error: Unsupported C feature (BDD-analysis does not support arrays: stack[__CPAchecker_TMP_1]) (BDDVectorCExpressionVisitor.visit, SEVERE)
"pthread-atomic/qrcu_false-unreach-call.i",           # OK!
"pthread-ext/28_buggy_simple_loop1_vf_false-unreach-call.i",      # PROBLEM! Verification result: UNKNOWN, incomplete analysis. Error: Unrecognized code (multiple thread assignments to same LHS not supported: t) (ThreadingTransferRelation.getNewThreadId, SEVERE)
```
We should re-evaluate all benchmarks before this commit is merged to the main branch.

**Comment by marek-trtik** [comment]
The fix/improvement of the `\result` guard condition is here:
https://github.com/marek-trtik/cbmc/commit/9bdb376a988e9ed8a406169b9ef79536813c8fc2
Please, look at it, and if you have any suspicion that it might cause troubles for some benchmark, then comment here. Thanks.

**Comment by tautschnig** [comment]
Looks ok to me!

**Comment by tautschnig** [comment]
> We should re-evaluate all benchmarks before this commit is merged to the main branch.

Do you have some reasonable way of doing so?

**Comment by marek-trtik** [comment]
Yes, I am in the progress of writing a Python script which will take all the benchmarks of that category, and for each of them it will call `benchexec` with 30s timeout and then CPAchecker to approve the witness. Hopefully it will work.

In this evaluation I'd like to use all today fixed we have (those to be merged to the main branch).

**Comment by tautschnig** [comment]
In an attempt to rule out at least one possible cause of trouble, I've added 476e65d6f7 to output `0` and `1` instead of `FALSE`/`TRUE`.

Yet I'm not sure how much of a difference that makes.
